### DISCUSSION TITLE
can all DUM methods be called concurrently by different threads on different DUM objects?

### DISCUSSION THREAD

**Discussion by Jacob-Burckhardt** [resiprocate/resiprocate#270] [Q&A] [answered]
[This link](https://github.com/resiprocate/resiprocate/wiki/DUM-Threading) says:

> NOTE: There is no threading protection in the DUM methods themseleves. If your application itself is multi-threaded you MUST provide your own protection (mutexing) if you plan on calling dum->process and/or other DUM API's from different threads.

Does DUM have similar rules that C++ standard library containers have?  For example, https://en.cppreference.com/w/cpp/container says:

> All container functions can be called concurrently by different threads on different containers.

Can all DUM methods be called concurrently by different threads on _different_ DUM objects?

Is the same true for SipStack?

**Comment by sgodin** [comment]
I'm going to assume that by DUM objects you mean different objects within DUM and not multiple instances of the DialogUsageManager object itself.

No you cannot call DUM API's from any thread.  You must either call all DUM API's from one thread (except DiagogUsageManager::post), or use a mutexed solution as described on the link: https://github.com/resiprocate/resiprocate/wiki/DUM-Threading#Tips_on_Making_Thread_Safe_Calls_when_using_DUM

As for SipStack, most API's are thread safe to call from any thread.  There are likely a few API's that aren't thread safe after you have started the stack processing loop, but unfortunately they are not clearly labelled.  You may need to examine the implementation code to know for sure.

**Reply by Jacob-Burckhardt** [reply]
I mean multiple instances of the DialogUsageManager object.  For example, this code creates ten threads and each thread has its own 
DialogUsageManager instance.  For each DUM instance, there is only one thread that accesses it.  Is that thread-safe?

```
   vector<thread> threads;
   for (int i = 0; i < 10; i++) {
      threads.push_back(thread{[]() {
         SipStack stackUa;
         DialogUsageManager dumUa{stackUa};
         while (true) {
            stackUa.process(50);
            while(dumUa.process());
         }
      }});
   }
```

**Comment by sgodin** [comment, marked as answer]
Yes that is thread safe.  :)On Oct 12, 2022, at 6:41 PM, Jacob-Burckhardt ***@***.***> wrote:﻿
I mean multiple instances of the DialogUsageManager object.  For example, this code creates ten threads and each thread has its own
DialogUsageManager instance.  For each DUM instance, there is only one thread that accesses it.  Is that thread-safe?
   vector<thread> threads;
   for (int i = 0; i < 10; i++) {
      threads.push_back(thread{[]() {
         SipStack stackUa;
         DialogUsageManager dumUa{stackUa};
         while (true) {
            stackUa.process(50);
            while(dumUa.process());
         }
      }});
   }

**Reply by Jacob-Burckhardt** [reply]
Thanks!
